### PR TITLE
Add read function for form file in CPLHTTPFetch

### PR DESCRIPTION
**Pull request by BishopGIS**
## What does this PR do?
Add custom read file function to curl in CPLHTTPFetch. 
On Windows curl needs file path in system encoding. So the path encoding in FORM_FILE_PATH variable differs from GDAL usual path encoding - UTF8.

### CODE REVIEW

File: gdal/port/cpl_http.cpp [hunk 1]
```diff
@@ -422,6 +422,33 @@ static size_t CPLHdrWriteFct( void *buffer, size_t size, size_t nmemb,
     return nmemb;
 }
 
+/************************************************************************/
+/*                          CPLReadFunction()                           */
+/************************************************************************/
+static size_t CPLReadFunction(char *buffer, size_t size, size_t nitems, void *arg)
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:428]: rename to CPLHTTPReadFunction

File: gdal/port/cpl_http.cpp [hunk 2]
```diff
@@ -422,6 +422,33 @@ static size_t CPLHdrWriteFct( void *buffer, size_t size, size_t nmemb,
...
+{
+    return VSIFReadL(buffer, size, nitems, static_cast<FILE*>(arg));
+}
+
+/************************************************************************/
+/*                          CPLSeekFunction()                           */
+/************************************************************************/
+static int CPLSeekFunction(void *arg, curl_off_t offset, int origin)
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:436]: rename to CPLHTTPSeekFunction

File: gdal/port/cpl_http.cpp [hunk 3]
```diff
@@ -422,6 +422,33 @@ static size_t CPLHdrWriteFct( void *buffer, size_t size, size_t nmemb,
     return nmemb;
 }
 
+/************************************************************************/
+/*                          CPLReadFunction()                           */
+/************************************************************************/
+static size_t CPLReadFunction(char *buffer, size_t size, size_t nitems, void *arg)
+{
+    return VSIFReadL(buffer, size, nitems, static_cast<FILE*>(arg));
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:430]:
```suggestion
    return VSIFReadL(buffer, size, nitems, static_cast<VSILFILE*>(arg));
```

File: gdal/port/cpl_http.cpp [hunk 4]
```diff
@@ -422,6 +422,33 @@ static size_t CPLHdrWriteFct( void *buffer, size_t size, size_t nmemb,
...
+}
+
+/************************************************************************/
+/*                          CPLSeekFunction()                           */
+/************************************************************************/
+static int CPLSeekFunction(void *arg, curl_off_t offset, int origin)
+{
+    if( VSIFSeekL( static_cast<FILE*>(arg), offset, origin ) == 0 )
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:438]:
```suggestion
    if( VSIFSeekL( static_cast<VSILFILE*>(arg), offset, origin ) == 0 )
```

File: gdal/port/cpl_http.cpp [hunk 5]
```diff
@@ -422,6 +422,33 @@ static size_t CPLHdrWriteFct( void *buffer, size_t size, size_t nmemb,
...
+        return CURL_SEEKFUNC_OK;
+    else
+        return CURL_SEEKFUNC_FAIL;
+}
+
+/************************************************************************/
+/*                          CPLFreeFunction()                           */
+/************************************************************************/
+void CPLFreeFunction(void *arg)
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:447]: add static 
+
rename to CPLHTTPFreeFunction

File: gdal/port/cpl_http.cpp [hunk 6]
```diff
@@ -422,6 +422,33 @@ static size_t CPLHdrWriteFct( void *buffer, size_t size, size_t nmemb,
...
+{
+    VSIFCloseL(static_cast<FILE*>(arg));
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:449]:
```suggestion
    VSIFCloseL(static_cast<VSILFILE*>(arg));
```

File: gdal/port/cpl_http.cpp [hunk 7]
```diff
@@ -933,21 +960,34 @@ CPLHTTPResult *CPLHTTPFetchEx( const char *pszURL, CSLConstList papszOptions,
 #if LIBCURL_VERSION_NUM >= 0x073800 /* 7.56.0 */
         mime = curl_mime_init(http_handle);
         curl_mimepart *mimepart = curl_mime_addpart(mime);
+        VSILFILE *mime_fp = nullptr;
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:963]: declaration should be moved to where it is first affected

File: gdal/port/cpl_http.cpp [hunk 8]
```diff
@@ -933,21 +960,34 @@ CPLHTTPResult *CPLHTTPFetchEx( const char *pszURL, CSLConstList papszOptions,
...
-            curl_mime_filedata(mimepart, pszFormFilePath);
+            VSIStatBufL sStat;
+            if( VSIStatL( pszFormFilePath, &sStat ) == 0)
+            {
+                mime_fp = VSIFOpenL( pszFormFilePath, "rb" );
+                if( mime_fp != nullptr )
+                {
+                    curl_mime_name(mimepart, pszFormFileName);
+                    curl_mime_filename(mimepart, pszFilename);
+            /// curl_mime_filedata(mimepart, pszFormFilePath);
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:986]:
```suggestion
```

**Review by rouault — CHANGES_REQUESTED** [review]

File: gdal/port/cpl_http.cpp [hunk 9]
```diff
@@ -940,14 +969,25 @@ CPLHTTPResult *CPLHTTPFetchEx( const char *pszURL, CSLConstList papszOptions,
...
+                pszFormFileName = pszFilename;
             }
 
 #if LIBCURL_VERSION_NUM >= 0x073800 /* 7.56.0 */
-            curl_mime_name(mimepart, pszFormFileName);
-            curl_mime_filedata(mimepart, pszFormFilePath);
+            VSIStatBufL sStat;
+            if( VSIStatL( pszFormFilePath, &sStat ) == 0)
+            {
+                VSILFILE *mime_fp = VSIFOpenL( pszFormFilePath, "rb" );
```
Comment (rouault) [inline gdal/port/cpl_http.cpp:982]: is it intended that no error is emitted when then stat() or open() fail ?

Comment (BishopGIS) [reply gdal/port/cpl_http.cpp:982]: There is no problem to send form without file. But it seems to me that check file exists is reasonable. I'll add checks for this case.

**Pull request merged**